numpy/core/tests/test_simd.py

```diff
@@ -0,0 +1,422 @@
+# NOTE: Please avoid the use of numpy.testing since NPYV intrinsics
+# may be involved in their functionality.
+import pytest
+from numpy.core._simd import targets
+
+class _Test_Utility:
+    # submodule of the desired SIMD extention, e.g. targets["AVX512F"]
+    npyv = None
+    # the current data type suffix e.g. 's8'
+    sfx  = None
+
+    def __getattr__(self, attr):
+        """
+        To call NPV intrinsics without the prefix 'npyv_' and
+        auto suffixing intrinsics according to class attribute 'sfx'
+        """
+        return getattr(self.npyv, attr + "_" + self.sfx)
+
+    def _data(self, n=None, reverse=False):
+        """
+        Create list of consecutive numbers according to number of vector's lanes.
+        """
+        if n is None:
+            n = 1
+        rng = range(n, n + self.nlanes)
+        if reverse:
+            rng = reversed(rng)
+        if self._is_fp():
+            return [x / 1.0 for x in rng]
+        return list(rng)
+
+    def _is_unsigned(self):
+        return self.sfx[0] == 'u'
+
+    def _is_signed(self):
+        return self.sfx[0] == 's'
+
+    def _is_fp(self):
+        return self.sfx[0] == 'f'
+
+    def _scalar_size(self):
+        return int(self.sfx[1:])
+
+    def _int_clip(self, seq):
+        if self._is_fp():
+            return seq
+        max_int = self._int_max()
+        min_int = self._int_min()
+        return [min(max(v, min_int), max_int) for v in seq]
+
+    def _int_max(self):
+        if self._is_fp():
+            return None
+        max_u = self._to_unsigned(self.setall(-1))[0]
+        if self._is_signed():
+            return max_u // 2
+        return max_u
+
+    def _int_min(self):
+        if self._is_fp():
+            return None
+        if self._is_unsigned():
+            return 0
+        return -(self._int_max() + 1)
+
+    def _true_mask(self):
+        max_unsig = getattr(self.npyv, "setall_u" + self.sfx[1:])(-1)
+        return max_unsig[0]
+
+    def _to_unsigned(self, vector):
+        if isinstance(vector, (list, tuple)):
+            return getattr(self.npyv, "load_u" + self.sfx[1:])(vector)
+        else:
+            sfx = vector.__name__.replace("npyv_", "")
+            if sfx[0] == "b":
+                cvt_intrin = "cvt_u{0}_b{0}"
+            else:
+                cvt_intrin = "reinterpret_u{0}_{1}"
+            return getattr(self.npyv, cvt_intrin.format(sfx[1:], sfx))(vector)
+
+class _SIMD_INT(_Test_Utility):
+    """
+    To test all integer vector types at once
+    """
+    def test_operators_shift(self):
+        if self.sfx in ("u8", "s8"):
+            return
```

Comment:
Instead of the `globals()` manipulation below, you could do something like
```suggestion
@pytest.mark.parametrize('sfx', ("u8", "s8", "u16", "s16", "u32", "s32", "u64", "s64"))
class _SIMD_INT(_Test_Utility):
    """
    To test all integer vector types at once
    """
    def test_operators_shift(self, sfx):
        if sfx in ("u8", "s8"):
            return
```

Reply:
Ah ok, here's the trick:

* Make the `SuffixInfo` class I suggest below
* Change these classes to use `sfx` as an argument instead of `self.sfx`
* Instead of the `globals()` mangling, do
```python
for sfxes, cls in tests_registry.items():
    params = []
    for target_name, npyv in targets.items():
        simd_width = npyv.simd if npyv else ''
        pretty_name = target_name.split('__') # multi-target separator
        if len(pretty_name) > 1:
            # multi-target
            pretty_name = f"({' '.join(pretty_name)})"
        else:
            pretty_name = pretty_name[0]

        mark = None
        skip_sfx = dict()

        for sfx in sfxes:
            if not npyv:
                mark = pytest.mark.skip(f"target '{pretty_name}' isn't supported by current machine")
            elif not npyv.simd:
                mark = pytest.mark.skip(f"target '{pretty_name}' isn't supported by NPYV")
            elif sfx == "f64" and not npyv.simd_f64:
                mark = pytest.mark.skip(f"target '{pretty_name}' doesn't support double-precision")
            else:
                mark = None

            params += [pytest.param(SuffixInfo(sfx=sfx, npyv=npyv, marks=mark, id=pretty_name))]

    pytest.mark.parametrize('sfx', params)(cls)
```

Reply:
I tried to use `pytest.mark.parametrize ()` before and it was totaly anoying and it doesn't provide a clean traceback
as the current implmention does.

Reply:
What was unclean about the traceback?

Reply:
that an example of how it looks when I use 
<details>
<summary>`pytest.mark.parametrize()`</summary>

```Bash
_____________________________________________________________________________________ TestSIMD_INT2.test_memory_load[npyv23] ______________________________________________________________________________________

self = <numpy.core.tests.test_simd.TestSIMD_INT2 object at 0x7fb33ec08e10>, npyv = <module 'NPYV_BASELINE'>

    def test_memory_load(self, npyv):
        # unaligned load
        load_data = npyv.load(npyv._data())
>       assert load_data == 0
E       assert <npyv_s64 of [1, 2]> == 0

load_data  = <npyv_s64 of [1, 2]>
npyv       = <module 'NPYV_BASELINE'>
self       = <numpy.core.tests.test_simd.TestSIMD_INT2 object at 0x7fb33ec08e10>

numpy/core/tests/test_simd.py:408: AssertionError
============================================================================================= short test summary info =============================================================================================
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv0] - assert <npyv_u8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30...
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv1] - assert <npyv_s8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30...
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv2] - assert <npyv_u16 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv3] - assert <npyv_s16 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv4] - assert <npyv_u32 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv5] - assert <npyv_s32 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv6] - assert <npyv_u64 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv7] - assert <npyv_s64 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv8] - assert <npyv_u8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv9] - assert <npyv_s8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv10] - assert <npyv_u16 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv11] - assert <npyv_s16 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv12] - assert <npyv_u32 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv13] - assert <npyv_s32 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv14] - assert <npyv_u64 of [1, 2]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv15] - assert <npyv_s64 of [1, 2]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv16] - assert <npyv_u8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv17] - assert <npyv_s8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv18] - assert <npyv_u16 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv19] - assert <npyv_s16 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv20] - assert <npyv_u32 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv21] - assert <npyv_s32 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv22] - assert <npyv_u64 of [1, 2]> == 0
FAILED numpy/core/tests/test_simd.py::TestSIMD_INT2::test_memory_load[npyv23] - assert <npyv_s64 of [1, 2]> == 0
````
</details>


<details>
<summary>the current implementation</summary>

```Bash
________________________________________________________________________________ Test_SIMD_INT2_128_baseline_u64.test_memory_load _________________________________________________________________________________

self = <numpy.core.tests.test_simd.Test_SIMD_INT2_128_baseline_u64 object at 0x7f897486d240>

    def test_memory_load(self):
        # unaligned load
        load_data = self.load(self._data())
>       assert load_data == 0
E       assert <npyv_u64 of [1, 2]> == 0

load_data  = <npyv_u64 of [1, 2]>
self       = <numpy.core.tests.test_simd.Test_SIMD_INT2_128_baseline_u64 object at 0x7f897486d240>

numpy/core/tests/test_simd.py:408: AssertionError
________________________________________________________________________________ Test_SIMD_INT2_128_baseline_s64.test_memory_load _________________________________________________________________________________

self = <numpy.core.tests.test_simd.Test_SIMD_INT2_128_baseline_s64 object at 0x7f8974840c50>

    def test_memory_load(self):
        # unaligned load
        load_data = self.load(self._data())
>       assert load_data == 0
E       assert <npyv_s64 of [1, 2]> == 0

load_data  = <npyv_s64 of [1, 2]>
self       = <numpy.core.tests.test_simd.Test_SIMD_INT2_128_baseline_s64 object at 0x7f8974840c50>

numpy/core/tests/test_simd.py:408: AssertionError
============================================================================================= short test summary info =============================================================================================
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_256_FMA3__AVX2_u8::test_memory_load - assert <npyv_u8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27...
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_256_FMA3__AVX2_s8::test_memory_load - assert <npyv_s8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27...
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_256_FMA3__AVX2_u16::test_memory_load - assert <npyv_u16 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_256_FMA3__AVX2_s16::test_memory_load - assert <npyv_s16 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_256_FMA3__AVX2_u32::test_memory_load - assert <npyv_u32 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_256_FMA3__AVX2_s32::test_memory_load - assert <npyv_s32 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_256_FMA3__AVX2_u64::test_memory_load - assert <npyv_u64 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_256_FMA3__AVX2_s64::test_memory_load - assert <npyv_s64 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_SSE42_u8::test_memory_load - assert <npyv_u8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_SSE42_s8::test_memory_load - assert <npyv_s8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_SSE42_u16::test_memory_load - assert <npyv_u16 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_SSE42_s16::test_memory_load - assert <npyv_s16 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_SSE42_u32::test_memory_load - assert <npyv_u32 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_SSE42_s32::test_memory_load - assert <npyv_s32 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_SSE42_u64::test_memory_load - assert <npyv_u64 of [1, 2]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_SSE42_s64::test_memory_load - assert <npyv_s64 of [1, 2]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_baseline_u8::test_memory_load - assert <npyv_u8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_baseline_s8::test_memory_load - assert <npyv_s8 of [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_baseline_u16::test_memory_load - assert <npyv_u16 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_baseline_s16::test_memory_load - assert <npyv_s16 of [1, 2, 3, 4, 5, 6, 7, 8]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_baseline_u32::test_memory_load - assert <npyv_u32 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_baseline_s32::test_memory_load - assert <npyv_s32 of [1, 2, 3, 4]> == 0
FAILED numpy/core/tests/test_simd.py::Test_SIMD_INT2_128_baseline_u64::test_memory_load - assert <npyv_u64 of [1, 2]> == 0

````
</details>

Reply:
> What was unclean about the traceback?

the summary

Reply:
Thanks for the log outputs. The trick is the use of `pytest.param(x,  id="some text")` instead of `x` in the list of values, which overrides what appears in the `[]`.

Reply:
Aaaah nice trick, now my second issue was the ability to override certain function in utility(`_Test_Utility`) which made me considered "annoying", for example:
```Python
class _SIMD_BOOL(_Test_Utility):
    """
    To test all boolean vector types at once
    """
    # override data generator without the need to introduce a new class 
    def _data(self, n=None, reverse=False):
        nlanes = getattr(self.npyv, "nlanes_u" + self.sfx[1:])
        true_mask = self._true_mask()
        rng = range(nlanes)
        if reverse:
            rng = reversed(rng)
        return [true_mask if x % 2 else 0 for x in rng]

    def _load(self, data):
        len_str = self.sfx[1:]
        load = getattr(self.npyv, "load_u" + len_str)
        cvt = getattr(self.npyv, f"cvt_b{len_str}_u{len_str}")
        return cvt(load(data))

    def test_tobits(self):
        if self.sfx in ("b8", "b16"):
            return
        for data in (self._data(), self._data(reverse=True)):
            vdata = self._load(data)
            data_bits = sum([x & (1 << i) for i, x in enumerate(data, 0)])
            tobits = self.tobits(vdata)
            assert tobits == data_bits
````

Reply:
@mattip, I need to hear your opinion here, since it seems I gonna have to follow @eric-wieser suggestion.

Reply:
If @mattip thinks we should just put this in as is, then I won't object.

Reply:
We could leave it for a follow-on clean up. As new intrinsics are added the tests will expand anyway. @seiko2plus if you choose to go that route please open an issue to review the tests here with a summary of the comments. This will also allow other people to contribute by refactoring the tests.

Reply:
The current implementation looks fine to me since using over `parametrize()` instead of the inheritance technique reduces the flexiblity, therfore I will keep the current route.
I belieive the current status of this work desreves the purple color, I don't mind going to second round of discussion but that
may lead to slow down the SIMD progress.

Reply:
I opened an issue gh-17667